sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,225 @@
+                );
+            }).onErrorResume((error) -> {
+                logger.debug("cache[{}] resulted in error", key, error);
+                if (initialLazyValue.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.empty();
+            });
+        }
+
+        logger.debug("cache[{}] doesn't exist, computing new value", key);
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+        return result.getValueAsync().onErrorResume(
+            (error) -> {
+                logger.debug("cache[{}] resulted in error", key, error);
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                this.remove(key);
+                return Mono.empty();
+            }
+        );
+    }
+
+    public void set(TKey key, TValue value) {
+        logger.debug("set cache[{}]={}", key, value);
+        AsyncLazyWithRefresh<TValue> updatedValue = new AsyncLazyWithRefresh<TValue>(value);
+        this.values.put(key, updatedValue);
+    }
+
+    public void remove(TKey key) {
+        values.remove(key);
+    }
+
+    /**
+     * This is AsyncLazy that has an additional Task that can
+     * be used to update the value. This allows concurrent requests
+     * to use the stale value while the refresh is occurring.
+     */
+    private class AsyncLazyWithRefresh<TValue> {
+        private final Function<TValue, Mono<TValue>> createValueFunc;
+        private final ReentrantLock valueLock = new ReentrantLock();
+        private final ReentrantLock removeFromCacheLock = new ReentrantLock();
+
+        private boolean removeFromCache = false;
+        private Mono<TValue> value;
+        private Mono<TValue> refreshInProgress;
+
+        public AsyncLazyWithRefresh(TValue value) {
+            this.createValueFunc = null;
+            this.value = Mono.just(value);
+            this.refreshInProgress = null;
+        }
+
+        public AsyncLazyWithRefresh(Mono<TValue> value) {
+            this.createValueFunc = null;
+            this.value = value;
+            this.refreshInProgress = null;
+        }
+
+        public AsyncLazyWithRefresh(Function<TValue, Mono<TValue>> taskFactory) {
+            this.createValueFunc = taskFactory;
+            this.value = null;
+            this.refreshInProgress = null;
+        }
+
+        public Mono<TValue> getValueAsync() {
+            Mono<TValue> valueMono = this.value;
+            if (valueMono != null) {
+                return valueMono;
+            }
+
+            valueLock.lock();
+            try {
+                if (this.value != null) {
+                    return this.value;
+                }
+
+                this.value = this.createValueFunc.apply(null);
```

Comment:
I think we probably need to cache the result here, turning into a hot publisher, else for each subscribe, it will trigger a call to gateway? 

https://github.com/Azure/azure-sdk-for-java/blob/d05f7e4ec323f32df36545871ef0685e4a645537/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncLazy.java#L38

https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#cache--

Reply:
You are correct, fixed it.